src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.cs

```diff
@@ -257,11 +233,68 @@ private void DrainQueue()
             }
         }
 
-        private static Solution MergeChanges(Solution solution, Solution? mutatedSolution)
+        private async Task<RequestContext> CreateRequestContextAsync(QueueItem queueItem, CancellationToken cancellationToken)
         {
-            // TODO: Merge in changes to the solution that have been received from didChange LSP methods
-            // https://github.com/dotnet/roslyn/issues/45427
-            return mutatedSolution ?? solution;
+            // This method asks the solution provider for a solution, and then updates it based on the state of the world
+            // as we know it. You may look at this and think "why doesn't the solution provider keep track of the state
+            // then it can just provide the right solutions" and at first glance that is appealing, but there are two big
+            // benefits to this queue tracking documents separately:
+            //
+            // 1. Since this queue manages the execution of handlers, we can ensure that the document state tracker
+            //    is only used from mutating request handlers, or outside of any handlers, and therefore we know
+            //    that calls to it cannot overlap, and it doesn't have to worry about threading.
+            //
+            // 2. We specifically only want the updated solution at the start of a request. If the solution provider
+            //    handed out "the right" solution, then a request could ask it for something at the start of processing
+            //    and at the end of processing, and get different results each time.
+
+            // There are multiple possible solutions that we could be interested in, so we need to find the document
+            // first and then get the solution from there. If we're not given a document, this will return the default
+            // solution
+            var (document, solution) = _solutionProvider.GetDocumentAndSolution(queueItem.TextDocument, queueItem.ClientName);
+
+            // Now we can update the solution to represent the LSP view of the world, with any text changes we received
+            solution = await GetSolutionWithReplacedDocumentsAsync(solution, cancellationToken).ConfigureAwait(false);
```

Comment:
We always need to do this because every request needs to act on a solution with the latest document text. The only difference between mutating and non-mutating from the document tracker point of view is that non-mutating isn't allowed to mutate document text. We enforce that by not passing it in.

Reply:
But the mutating operations (open/close/change text) aren't actually looking at the textual contents, indeed their goal is to _not_ look at them but instead record the text coming in.

Reply:
Change definitely needs the text, as it gets incremental changes. Open and close don't, but that feels like implementation details to me.

Are you thinking we break down the request types further to avoid this work for some requests? That seems like a risky thing to do because a bad handler that doesn't declare it's intentions correctly can always just party on the workspace and act on stale data. Having said that I'm not familiar enough with the perf characteristics enough to understand that risk.